### ISSUE TITLE
Category Sort Order is not saved in Flat Table

### ISSUE THREAD

**Issue by stanislav-zgr** [magento/magento2#3993]
## Preconditions
Magento 2

## Steps to reproduce
1. Enable Flat Catalog
2. Change the order of subcategories within one parent category (by dragging them). 

## Expected result
1. Position value is updated in Flat table for all affected subcategories of this level (where position has changed)

## Actual result
1. Only the subcategory which has been replaced is updated (plus parent categories)
2. The order of categories on the frontend becomes incorrect until Flat Category is re-indexed

**Comment by duhon** [comment]
Hello @stanislav-zgr 
Thank you for the report! We have created issue in our bugtracker system MAGETWO-54060

**Comment by azeemism** [comment]
@stanislav-zgr Thanks for posting this issue. I was really struggling trying to sort categories that just didn't transfer over to the website category menu. I am using 2.1.2 CE. @duhon @vkorotun hopefully some is assigned this issue.
Thanks.

**Comment by VincentMarmiesse** [comment]
Hello,

Any news on this issue?

Thanks!

**Comment by magento-engcom-team** [comment]
@stanislav-zgr, thank you for your report.
We were not able to reproduce this issue by following the steps you provided. If you'd like to update it, please reopen the issue.
We tested the issue on 2.3.0-dev, 2.2.0, 2.1.9

**Comment by josephmcdermott** [comment]
This issue still occurs on Magento 2.2.5. It is very easy to replicate, following the original poster's instructions. Flat catalog on, drag a category to a new position, and check the Flat Category table.

Only the category updated has its position updated, so you end up with a number of categories with the same position. The only remedy currently is a full flat catalog category reindex, which is obviously overkill as this should be handled automatically.

In this example I moved around Category 4 and Category 5, which is reflected in the Admin panel (which uses the EAV value) but not on the frontend (which uses the FLAT value):

```
mysql> select "eav" as dbtable,ccev.value, cce.position from catalog_category_entity cce inner join catalog_category_entity_varchar ccev on cce.entity_id = ccev.entity_id where ccev.attribute_id = 45 and cce.level = 2 order by position asc;
+---------+--------------------+----------+
| dbtable | value              | position |
+---------+--------------------+----------+
| eav     | Category 1         |        1 |
| eav     | Category 2         |        2 |
| eav     | Category 3         |        3 |
| eav     | Category 5         |        4 |
| eav     | Category 4         |        5 |
| eav     | Category 6         |        6 |
+---------+--------------------+----------+

mysql > select "flat" as dbtable, name, position from catalog_category_flat_store_1 where level = 2 order by position asc;
+---------+--------------------+----------+
| dbtable | name               | position |
+---------+--------------------+----------+
| flat    | Category 1         |        1 |
| flat    | Category 2         |        2 |
| flat    | Category 3         |        3 |
| flat    | Category 4         |        3 |
| flat    | Category 5         |        3 |
| flat    | Category 6         |        6 |
+---------+--------------------+----------+
```

**Comment by chris-pook** [comment]
Re-opening as this is still an issue as per the comment above.

**Comment by josephmcdermott** [comment]
Could it be that this line needs to retrieve all sibling category ID's of the category being moved?
https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Catalog/Model/Category.php#L460

eg:
```php
if (!$flatIndexer->isScheduled()) {
    $categoryIdsToReindex = array_merge(
        [$this->getId(), $oldParentId, $parentId],
        $this->getResource()->getChildren($this->categoryRepository->get($parentId), false),
        $this->getResource()->getChildren($this->categoryRepository->get($oldParentId), false)
    );
    $flatIndexer->reindexList($categoryIdsToReindex);
}
```

This way the category, its previous siblings and new siblings all get their position reindexed.

Note: the `getChildren` method for retrieving children IDs isn't perfect, as it ignores is_active: 0 categories.

**Comment by ghost** [comment]
@stanislav-zgr, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by magento-engcom-team** [comment]
Hi @stanislav-zgr. Thank you for your report. 
The issue has been fixed in magento/magento2#23820 by @sergey-solo in 2.3-develop branch
Related commit(s):
 - [089bebcf1ef29b13e61cf696d63cf346cb274d56](https://github.com/magento/magento2/commit/089bebcf1ef29b13e61cf696d63cf346cb274d56)
 - [592922ea34707ad8cadb92f26d967ebd0db03e30](https://github.com/magento/magento2/commit/592922ea34707ad8cadb92f26d967ebd0db03e30)

The fix will be available with the upcoming 2.3.4 release.
